Rector 0.10

mglaman opened the pull request:
Fixes #133
Fixes #135

## Remaining items

- [ ] Find out why parsing `src/Plugin/WebformElement/WebformLikert` crashed PHPStan's doc parser (This should be a follow up, it's not a Rector or this package bug)
- [x] Do not assume PHPUnit 7 in trait compatibility 

## Required follow ups

These are issues that need a fast follows once we merge this PR, otherwise the PR will just grow and grow and grow.

-  #138 Constants defined in .module files are not discovered
- Drupal 9 does not have PhpUnitCompatibility test traits (see https://github.com/palantirnet/drupal-rector/pull/134#issuecomment-821551267)
- Messaging that users should upgrade to 8.9 if on 8.7, 8.8, etc (see https://github.com/palantirnet/drupal-rector/pull/134#issuecomment-821547909)

## How to test

Add fork to your Drupal **8** site (assuming Composer build) composer.json
```
cd /path/to/drupal/recommended-project
composer config repositories.1 vcs https://github.com/bluehorndigital/drupal-rector.git
```

Require it
```
composer require palantirnet/drupal-rector:dev-rector-0.10
```

Configure it
```
cp vendor/palantirnet/drupal-rector/rector.php .
```

Run it!

Test it on your own code

```
php vendor/bin/rector process web/modules/custom
```

**OR** Grab the examples
```
cp -r vendor/palantirnet/drupal-rector/rector_examples web/modules/
php vendor/bin/rector process web/modules/rector_examples --dry-run
```

Verify it!

## Notes
Takes #122 commits and updates to 0.10

- composer: update rector to 0.8
- switch configs from YML to PHP
- drop duplicated info in editorconfig
- Add phpstan and rector dev trunk
- Add back EntityDeleteMultipleRector
- Update rule definitions for Rector 0.10

TomasVotruba committed:
composer: update rector to 0.8

TomasVotruba committed:
switch configs from YML to PHP

TomasVotruba committed:
drop duplicated info in editorconfig

mglaman committed:
Add phpstan and rector dev trunk

mglaman committed:
Merge branch 'rector-08' into rector-0.10

# Conflicts:
#	.editorconfig
#	composer.json
#	config/drupal-8/drupal-8.0-deprecations.yml

mglaman committed:
Add back EntityDeleteMultipleRector

mglaman committed:
Update rule definitions for Rector 0.10

mglaman committed:
Fix remaining usages of Option::EXCLUDE_PATHS

Also adds PHPStan workflow to catch unused calls.

mglaman committed:
PHPStan workflow to catch unused calls.

mglaman committed:
Add PHPUnit and initial test

mglaman committed:
Update local packages flow

mglaman committed:
Handle ~/drupal in specific workflow

mglaman committed:
Normalize GitHub flows

mglaman committed:
Add PHPUnit flow, move deprecation-index tests to PHPUnit

mglaman committed:
Move away from Behat and to specific commands

mglaman committed:
Must add php suffix

mglaman committed:
Run with debug, remove duplicated core discovery

mglaman committed:
Add missing 8.4 deprecation Rector rules config

mglaman committed:
Ensure we load the PHPUnit bootstrap for test class autoloading

mglaman committed:
Update the integration test

mglaman commented on tests/config/rector-phpunit.php:
```diff
@@ -0,0 +1,32 @@
+        __DIR__ . '/../fixtures/drupal/web/modules',
+        __DIR__ . '/../fixtures/drupal/web/profiles'
+    ]);
+    require_once __DIR__ . '/../fixtures/drupal/web/core/tests/bootstrap.php';
```
This causes a bug, see

```
PHP Fatal error:  Cannot declare class Symfony\Component\DependencyInjection\Compiler\MergeExtensionConfigurationParameterBag, because the name is already in use in /home/runner/work/drupal-rector/drupal-rector/tests/fixtures/drupal/vendor/symfony/dependency-injection/Compiler/MergeExtensionConfigurationPass.php on line 135
```

mglaman committed:
Expect dry-run exit code 1

mglaman committed:
Remove phpunit bootstrap

mglaman committed:
Update examples_updated for db_*

mglaman committed:
Fix comments: parentNode is now just parent

mglaman committed:
nextNode is just next

mglaman committed:
Fix remaining diff errors

mglaman committed:
rector/rector:0.10.4 stopped returning applied rectors

TomasVotruba commented on src/Utility/AddCommentTrait.php:
```diff
@@ -23,8 +23,8 @@ protected function getClosestStatementNode(Node $node): ?Node {
         }
-        elseif ($node->hasAttribute('parentNode')) {
-            $parent_node = $node->getAttribute('parentNode');
+        elseif ($node->hasAttribute('parent')) {
```
Btw, it's more future proof to use `AttributeKey::PARENT` constant

mglaman replied:
:O yes, I will work on that fix.

mglaman commented on rector.php:
```diff
@@ -0,0 +1,22 @@
+
+    $parameters = $containerConfigurator->parameters();
+    $parameters->set(Option::AUTOLOAD_PATHS, ['web/core', 'web/modules', 'web/profiles', 'web/themes']);
+    require_once __DIR__ . '/web/core/tests/bootstrap.php';
```
This fixes PHPUnit trait but kills any output generated by Rector.

mglaman commented on rector-config-web-dir.php:
```diff
@@ -0,0 +1,21 @@
+
+    $parameters = $containerConfigurator->parameters();
+    $parameters->set(Option::AUTOLOAD_PATHS, ['web/core', 'web/modules', 'web/profiles', 'web/themes']);
+    require_once __DIR__ . '/web/core/tests/bootstrap.php';
```
This fixes PHPUnit trait but kills any output generated by Rector.

mglaman commented on rector-config-docroot-dir.php:
```diff
@@ -0,0 +1,22 @@
+    $parameters = $containerConfigurator->parameters();
+
+    $parameters->set(Option::AUTOLOAD_PATHS, ['docroot/core', 'docroot/modules', 'docroot/profiles', 'docroot/themes']);
+    require_once __DIR__ . '/docroot/core/tests/bootstrap.php';
```
This fixes PHPUnit trait but kills any output generated by Rector.

mglaman reviewed (commented):
@damontgomery I found a bug in the PHPUnit test trait autoloading which kills all output from Rector :( I'll work on a fix.

mglaman committed:
Use constants for accessing attribute values

mglaman committed:
Remove include of "core/tests/bootstrap.php"

mglaman committed:
0.10.6 restores ability to check file_diffs for applied rules

jibran commented on README.md:
````diff
@@ -58,23 +58,23 @@ $ composer require --dev palantirnet/drupal-rector
 
 ```bash
-cp vendor/palantirnet/drupal-rector/rector-config-web-dir.yml rector.yml
+cp vendor/palantirnet/drupal-rector/rector-config-web-dir.yml rector.php
````
s/rector-config-web-dir.yml/rector-config-web-dir.php

jibran commented on README.md:
````diff
@@ -58,23 +58,23 @@ $ composer require --dev palantirnet/drupal-rector
 
 ```bash
-cp vendor/palantirnet/drupal-rector/rector-config-docroot-dir.yml rector.yml
+cp vendor/palantirnet/drupal-rector/rector-config-docroot-dir.yml rector.php
````
s/rector-config-docroot-dir.yml/rector-config-docroot-dir.php

jibran commented on README.md:
````diff
@@ -58,23 +58,23 @@ $ composer require --dev palantirnet/drupal-rector
 ```
 
-If your document root directory is something else you will need to manually copy and edit `rector.yml`.
+If your document root directory is something else you will need to manually copy and edit `rector.php`.
````
Replace the `web` in these paths with your document root.

```
parameters:
  autoload_paths:
    - 'web/core'
    - 'web/core/modules'
    - 'web/modules'
    - 'web/profiles'
```
This code block needs an update as it is still in yml.

mglaman replied:
Thank you :)

mglaman committed:
Update remaining `yml` documentation

mglaman committed:
Initial work to fix test traits autoloading

mglaman committed:
Make PHPUnit autoloading bootstrap file generic for all config

mglaman commented on config/drupal-phpunit-bootstrap-file.php:
```diff
@@ -0,0 +1,159 @@
+assert($this instanceof BootstrapFilesIncluder);
+/** @phpstan-ignore-next-line */
+$parameterProvider = $this->parameterProvider;
+assert($parameterProvider instanceof ParameterProvider);
```
@TomasVotruba how horrible of an atrocity am I committing here? We need the parameters in Rector to resolve the Drupal path for fixing some autoloading

TomasVotruba replied:
I see. Maybe it would be simpler to use `PhpFileLoader` from Symfony and load just `rector.php` via ContainerBuilder to get parameters.

Much cleaner than porting whole RectorFactory and building Rector again IMO

mglaman replied:
This is inside of a `bootstrap_file` which gets required in the main scope, so it's not porting and building rector.

See: https://github.com/rectorphp/rector/blob/06f85e4a68acba9ea05d7894e79df5661663e452/src/Autoloading/BootstrapFilesIncluder.php#L38

```
        $bootstrapFiles = $this->parameterProvider->provideArrayParameter(Option::BOOTSTRAP_FILES);

        foreach ($bootstrapFiles as $bootstrapFile) {
            if (! is_file($bootstrapFile)) {
                throw new ShouldNotHappenException('Bootstrap file %s does not exist.', $bootstrapFile);
            }

            try {
                require_once $bootstrapFile;
            } catch (Throwable $throwable) {
```

I know PHPStan actually wraps the include in a static function to restrict the scope, but also passes a few things via `use ()`.

I didn't know if there was an easier way to have a bootstrap file / autoloader which was aware of the current parameters (without injecting globals)

mglaman committed:
Use webflo/drupal-finder

mglaman committed:
Ensure test/fixtures directory

mglaman committed:
Update phpstan.neon to remove old config files

mglaman commented on composer.json:
```diff
@@ -28,10 +29,18 @@
         }
     ],
     "autoload": {
+        "files": [
+            "drupal-phpunit-hack.php"
+        ],
```
I can dig up the live stream where I worked on this for phpstan-drupal. It's not pretty. But I have no other idea how we could possibly mimic Drupal's runtime mutation. It creates a mutated class copy into `sites/simpletest` and essentially overwrites the default PHPUnit one by getting it autoloaded first.

mglaman commented on config/drupal-phpunit-bootstrap-file.php:
```diff
@@ -0,0 +1,168 @@
+<?php
+
+/**
+ * @file
+ *
+ * This fixes Drupal testing namespace autoloading and PHPUnit compatibility.
+ */
```
🤔 I should probably specify this is mostly a copy job of `core/tests/bootstrap.php`

damontgomery replied:
++ It's been a while and I'm not too familiar with all the specifics if we need updates down the road.

Can you comment on the places that were modified?

mglaman commented on config/drupal-phpunit-bootstrap-file.php:
```diff
@@ -0,0 +1,168 @@
+
+// Force the alias to PHPUnit 7. This prevents possibly aliasing to PhpUnit9.
+require_once $drupalRoot . '/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit7/TestCompatibilityTrait.php';
+class_alias("Drupal\TestTools\PhpUnitCompatibility\PhpUnit7\TestCompatibilityTrait", '\Drupal\Tests\PhpunitVersionDependentTestCompatibilityTrait');
```
I can try to explain this more.

mglaman commented on tests/fixtures/.gitignore:
```diff
@@ -0,0 +1 @@
+/MutatedTestCase.php
```
This ensures the fixtures directory exists, and that the weird mutated test class does not get commited.

mglaman commented on config/drupal-phpunit-bootstrap-file.php:
```diff
@@ -0,0 +1,168 @@
+assert($this instanceof BootstrapFilesIncluder);
+/** @phpstan-ignore-next-line */
+$parameterProvider = $this->parameterProvider;
+assert($parameterProvider instanceof ParameterProvider);
```
This assertion failed. It's probably due to rector-prefixed. But our workflows _should_ be mimicking that.

mglaman replied:
I verified the workflows are using rector-prefixed and the notice wasn't thrown. But, that's because the PHP environment isn't configured to catch assertions, I believe.

<img width="576" alt="Screen Shot 2021-04-16 at 8 08 16 AM" src="https://user-images.githubusercontent.com/3698644/115028807-e899c880-9e8a-11eb-8da1-aa780c5d2b98.png">

So instead of assert, we can use regular doc type hinting and an `if NULL`

mglaman replied:
Fixed with https://github.com/palantirnet/drupal-rector/pull/134/commits/4f498c8d8741b2f1661fdc01b4b3a9c20b4baf8c

KarinG replied:
Confirming Fixed

mglaman committed:
Fix notice on assert for a prefixed class

mglaman committed:
Run a test against webform for catching output errors

mglaman committed:
Dynamic PhpunitVersionDependentTestCompatibilityTrait version aliasing

mglaman committed:
Do not error with Webform if it fails

mglaman committed:
Run on different PHPUnit versions, fix hardcoded trait namespace

joeparsons reviewed (approved):
I think it would be great to merge this and address any remaining issues (e.g. Drupal <8.8 or 9+ compatibility) in separate issues/PRs.  Amazing work @mglaman!

damontgomery reviewed (commented):
Thanks again, @mglaman.

Testing steps I used:
- Use the Drupal Rector Sandbox (which uses DDEV)
- Add your branch within local /drupal-rector
- Update the composer dependencies, `composer update palantirnet/drupal-rector -W`
- Copy `rector.php` into the project root
- Run Drupal Rector, `vendor/bin/rector process web/modules/custom/rector_examples --dry-run`

(HAS SINCE BEEN RESOLVED) Got the error:
```
PHP Fatal error:  Declaration of RectorPrefix20210420\Symfony\Component\DependencyInjection\ServiceLocator::has(string $id) must be compatible with RectorPrefix20210420\Psr\Container\ContainerInterface::has($id) in /Users/dan/Sites/drupal-rector-sandbox/vendor/rector/rector-prefixed/vendor/symfony/dependency-injection/ServiceLocator.php on line 25
Fatal error: Declaration of RectorPrefix20210420\Symfony\Component\DependencyInjection\ServiceLocator::has(string $id) must be compatible with RectorPrefix20210420\Psr\Container\ContainerInterface::has($id) in /Users/dan/Sites/drupal-rector-sandbox/vendor/rector/rector-prefixed/vendor/symfony/dependency-injection/ServiceLocator.php on line 25
```

Going to keep looking over the code for readability, etc.

damontgomery commented on .github/workflows/local_package_functional_tests.yml:
```diff
@@ -14,46 +14,44 @@ jobs:
-                    # Run Behat
-                    vendor/bin/behat
+                  mv vendor/palantirnet/drupal-rector/rector_examples web/modules/custom
+                  ls -la web/modules/custom/rector_examples
```
Minor: Do we still need this? `ls -la`

mglaman replied:
Nope, we do not, it's debugging code

damontgomery commented on config/drupal-8/drupal-8-all-deprecations.php:
```diff
@@ -0,0 +1,16 @@
+    $containerConfigurator->import(__DIR__ . '/drupal-8.*');
+
+    $parameters = $containerConfigurator->parameters();
+    $parameters->set(Option::SKIP, []);
```
Developer experience: Can we comment this?

It isn't clear to me why we are setting something that seems to be a likely default. If that isn't the case, can we explain that?

mglaman replied:
We can, but with the new Rector there are no defaults. So you have to provide them, I'm fairly certain. I honestly don't know what this does, it was from the original upgrade PR

damontgomery replied:
@TomasVotruba, Do you know if this is needed and if so, what it does? We'd like to add a comment in the code as to why it's included.

TomasVotruba replied:
I would go with YAGNI and drop it :)

damontgomery replied:
Thanks, @TomasVotruba!

damontgomery commented on rector.php:
```diff
@@ -0,0 +1,31 @@
+    $parameters->set(Option::IMPORT_SHORT_CLASSES, false);
+    $parameters->set(Option::IMPORT_DOC_BLOCKS, false);
+
+    $parameters->set('drupal_rector_notices_as_comments', true);
```
Developer experience: Can we restore / replace comments?

People are expected to make small changes to this file and the comments may be helpful for common use cases.

mglaman replied:
@damontgomery actually this keeps comments. Commits like this fixed commenting, it was one of the first regressions found: https://github.com/palantirnet/drupal-rector/commit/02c49a54bb7a61ac97a20db9987358ca5501128e

damontgomery replied:
@mglaman, sorry, this was a bad place to add my question / suggestion. :(.

I meant the comments like `# This will add comments to call out edge cases in replacements.` in https://github.com/palantirnet/drupal-rector/blob/master/rector.yml.

damontgomery reviewed (commented):
WIP review. Have to jump on a call.

Overall this makes a lot of sense to me.

I left a few comments about developer experience.

I think the automated tests, Github Actions / Behat need to be reviewed. The Behat tests can likely be completely removed since they were replaced and the Github Actions jobs need to be consistent and we may be able to remove some of them if they are no longer needed.

I may be able to take a closer look at this later, but I wanted to post in-progress comments in case it's helpful.

Thanks again.

damontgomery committed:
Refactor Gitflow Actions.

Try to be clear about copy & paste steps.

We could share jobs, but these run faster.

Removed Behat mentions since we aren't using it.

Updated the local composer.json repository config to add names and split it up to be easier to read.

damontgomery committed:
Removed Behat automated tests.

These are included directly in Github Actions.

damontgomery committed:
Include Symfony YAML since it's required for the PhpUnit tests.

damontgomery committed:
Recursive copy for functional test.

mglaman committed:
Document running PHPUnit

mglaman committed:
Add documentation about PHP version
